### PR TITLE
Dockerize

### PR DESCRIPTION
**Pull request by ojoeth**

### CODE REVIEW

**Comment by cloudrac3r** [comment]
Hi, I'm not familiar with Docker. Do you think you could describe in detail what the script does, and give a short answer to why it is necessary? :)

**Comment by Perflyst** [comment]
> FROM node:alpine

Base image on which the application runs later (i think 12.13.0 should be tagged here as suggested in the readme)

> RUN apk add git python3 make g++

Installs additional packages required for doing stuff (git is not required here)

> WORKDIR /app

Where Bibliogram will be installed

> COPY . .

Copies the source code to the image

> RUN npm install --no-optional

Executes "npm install --no-optional" to install node modules

> EXPOSE 10407

Bibliogram runs on this port, it needs to be accessible

> CMD ["npm", "start"]

This gets executed on docker run, to start Bibliogram

**Comment by cloudrac3r** [comment]
Is /app relative to the root of the entire filesystem, or is it inside a docker sandbox of sorts?

12.18.0 is out now, if that's available, you should "tag" that.

What happens to the config file? Can it be modified by the person using the image?

What happens to the database file? Will it ever be erased?

If either of those things could be erased by some automated process related to Docker's operations, then they should be in a safe place and then linked or copied in for Docker to use.

Thank you for the summary.

**Comment by Perflyst** [comment]
> Is /app relative to the root of the entire filesystem, or is it inside a docker sandbox of sorts?

Inside the docker fs

> What happens to the config file? Can it be modified by the person using the image?

You need to "mount" it. `docker run -v ./config.js:/app/config.js -p 10407:10407 bibliogram`

> What happens to the database file? Will it ever be erased?

With this image it will be erased on restart. @ojoeth should either create a volume directly in the Dockerfile or the user needs to map it with a volume (like with the config above). I would prefer the second option, so that you can easily run "without" a database

**Comment by cloudrac3r** [comment]
Okay. The database obviously needs to be permanent before I'll accept this.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+FROM node:alpine
```
Comment (Perflyst) [inline Dockerfile:1]:
```suggestion
FROM 12.18.0-alpine
```

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+RUN apk add git python3 make g++
```
Comment (Perflyst) [inline Dockerfile:2]:
```suggestion
RUN apk add python3 make g++
```

**Comment by cloudrac3r** [comment]
@ojoeth Please read perflyst's review and suggestions, particularly the part about the database :)

If you need to know where the database files live, they're in `/db`. The main file is `/db/bibliogram.db`, the others are backups from previous versions.

**Comment by pratikbalar123** [comment]
add `VOLUME ["/db"]` in Dockerfile before the last line.

or add `-v /db:/db` in docker run command.

**Comment by giocomai** [comment]
I think there should also be a `.dockerignore`, with at least `node_modules` in it, otherwise the `node_modules` folder goes into the Docker build, and there are issues with modules compiled with a different node version. See e.g. this [SO question](https://stackoverflow.com/questions/57932529/linux-x64-binaries-cannot-be-used-on-the-linuxmusl-x64-platform-error) with reference to exactly the kind of error that currently comes out of the current Dockerfile with the sugested edits.

Or fix this in the Dockerfile, copying only what is needed.

**Comment by kalak-sudo** [comment]
GREAT solution, I hope soon will be working perfectly. 

I have just tried `docker pull giocomai/bibliogram` from: [https://hub.docker.com/r/giocomai/bibliogram](https://hub.docker.com/r/giocomai/bibliogram)

It was up and running but I didn't see anything at localhost:10407.

Hope it gets fixed soon, as docker is much easier to use than manual installing.

**Comment by giocomai** [comment]
A pull request is probably not the best place to discuss this, but to clarify how it works, this is how I set it up with docker-compose:

```
  bibliogram:
    image: giocomai/bibliogram:latest
    volumes:
      - ./bibliogram/config.js:/app/config.js:ro
      - bibliogram_db:/db
    ports:
      - "10407:10407"
    restart:
      unless-stopped
```
instead of docker-compose, you can achieve the same with docker passing the volumes with the  -v argument. 

A volume with the database is created automatically, and you don't really need to care about it. 

This however assumes that from the folder where you're starting in there will be a folder called `bibliogram` with a `config.js` in it (see repo for details). 

For reference, this is my whole dockerfile now:

```
FROM node:12.18.0-alpine
RUN apk --no-cache add git python3 make g++
WORKDIR /app
COPY . .
RUN npm install --no-optional
EXPOSE 10407
CMD ["npm", "start"]
```

and I have a .dockerignore with one line:
`node_modules`

This all works nicely for me.

**Comment by cloudrac3r** [comment]
Again, I have absolutely no experience with docker, but if someone sends me a quick tutorial on how I can set everything up with that dockerfile, I'll take a look and see if it works as intended.

**Comment by giocomai** [comment]
So in brief, if you want to build your own Docker image, in the main folder of your repo you create a file called `Dockerfile` and you put those lines above, one created `.dockerignore`, with a single line, `node_modules`. 

And then from the terminal from within that folder you run:

`docker build -t bibliogram .`

or

`docker build -t username/bibliogram .` (if you want to push it do Dockerhub).

Then you should be able to run it simply with: 
`docker run -p 10407:10407 giocomai/bibliogram`
or
`docker run -p 10407:10407 giocomai/bibliogram`

Or creating a volume with the -v argument to make sure the database is persistent among sessions:
`docker run -p 10407:10407 -v bibliogram_db:/db giocomai/bibliogram`

If you run this last command without doing any of the above, and don't need to customise config.js, this will automatically download from Docker Hub the container I built, and you'll find bibliogram up and running on localhost:10407. So that would be a one-liner to have bibliogram working without doing anything else.

If you want to build it yourself, just put your username there, or just `bibliogram` with nothing before the slash if you just called it that when you built it. 

Not nicely written, but I hope this helps. 

I have it running in the cloud, and it works nicely so far... if not for the fact that I need to run the curl thing to make it discover users. Locally, it runs perfectly.

**Comment by giocomai** [comment]
Also, to clarify, I put it on Docker Hub as I always leave everything public, but I suppose it would be best for the maintainer to publish it... after doing the build, with `docker build -t username/bibliogram .` you'd just need to push it to Docker Hub with `docker push username/bibliogram` (after setting up an account on Docker Hub), and there you go... you just run it when update other things in the repo, and people can just use the built dockerfile without worrying about the rest, and be up and running in a single line... if they trust the repo maintainer, it should be fine for them to use a Docker container built by them.

Also, for reference, as for example recent Fedora versions do not support easily docker, Fedora users should just put `podman` instead of `docker` in the commands above and have exactly the same result.

**Comment by cloudrac3r** [comment]
I should also mention that this repository does not use any version numbers, because every commit to master is fine for release and can be deployed everywhere. How does docker handle updates? Will they be downloaded automatically?

\-

> if not for the fact that I need to run the curl thing

@giocomai: https://github.com/cloudrac3r/bibliogram/wiki/Rate-limits#tldr-what-does-it-mean-if-an-instance-is-blocked

**Comment by kalak-sudo** [comment]
> Also, to clarify, I put it on Docker Hub as I always leave everything public, but I suppose it would be best for the maintainer to publish it... after doing the build, with `docker build -t username/bibliogram .` you'd just need to push it to Docker Hub with `docker push username/bibliogram` (after setting up an account on Docker Hub), and there you go... you just run it when update other things in the repo, and people can just use the built dockerfile without worrying about the rest, and be up and running in a single line... if they trust the repo maintainer, it should be fine for them to use a Docker container built by them.
> 
> Also, for reference, as for example recent Fedora versions do not support easily docker, Fedora users should just put `podman` instead of `docker` in the commands above and have exactly the same result.

### 

It would be great to put some details of the installation @giocomai at the docker hub repo.

**Comment by giocomai** [comment]
@cloudrac3r in brief, if you don't say a version number, each time you build (or push), the build is automatically tagged with "latest".

To update the new version, you'd just need to rebuild it again with:

`docker build -t username/bibliogram .`
and
`docker push username/bibliogram`

(adding the --no-cache param when building if you feel it's not picking up something, but it shouldn't be needed in your case)

these are exactly the same as, e.g., `docker build -t username/bibliogram:latest .` and `docker push username/bibliogram:latest`, as the `latest` tag is added by default.

If you had versions, you'd use: `docker build -t username/bibliogram:v2.3 .` (or whatever)

To update, the users would just run:

`docker pull username/bibliogram`, which would automatically get the latest version you updated. 

The next time they run docker, they would be using the updated verson. Hope this clarifies. 

@kalak-sudo ok, I'll add the description above on Docker hub, but this is just temporary until there is an "official" image... then I'll gladly just include a link to that one.

**Comment by cloudrac3r** [comment]
Okay. It would be really good if there was a way for users to be able to automatically apply updates, because I don't want to deal with issue reports from users who are using [instances that are over a month old](https://bibliogram.ggc-project.de/).

I managed to get docker and docker-compose working on my computer and I've been able to test your dockerfiles and change some things for myself. Seems like OP has vanished, or has no interest in continuing this, so I'll close this PR and commit the changes myself.

I will then open a new thread for us to move this docker discussion to.

Once I'm happy with the state of the dockerfiles I'll send them to docker hub under some official name and update the readme.

**Comment by cloudrac3r** [comment]
Let's go! I'll write a summary post of the outstanding comments from this PR so newcomers can get caught up. → https://github.com/cloudrac3r/bibliogram/issues/81

**Comment by giocomai** [comment]
sounds great. yes, users will have to just run `docker pull username/bibliogram` (as they would git pull from the main repo)... unless they need to run some update script on the database or something, which would then need additional instructions.

**Pull request closed without merging**